src/Conversion/ONNXToTorch/NN/ReduceMean.cpp

```diff
@@ -0,0 +1,142 @@
+#include "llvm/ADT/StringExtras.h"
+
+#include "torch-mlir/Dialect/TorchConversion/IR/TorchConversionDialect.h"
+#include "torch-mlir/Dialect/TorchConversion/IR/TorchConversionOps.h"
+#include "torch-mlir/Dialect/TorchConversion/Transforms/BackendTypeConversion.h"
+
+#ifdef _WIN32
+#include <io.h>
+#endif
+
+using namespace mlir;
+using namespace mlir::torch;
+using namespace mlir::torch::Torch;
+
+struct ONNXReduceMeanOpToTorchLowering : public ConversionPattern {
+public:
+  ONNXReduceMeanOpToTorchLowering(
+      TypeConverter &typeConverter, MLIRContext *ctx)
+      : ConversionPattern(typeConverter,
+            mlir::ONNXReduceMeanOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+
+    ONNXReduceMeanOp op1 = llvm::dyn_cast<ONNXReduceMeanOp>(op);
+    ONNXReduceMeanOpAdaptor adapter(op1);
+    mlir::MLIRContext *context = op1.getContext();
+    Location loc = op1.getLoc();
+
+    auto axes = op1.axesAttr();         // ::mlir::ArrayAttr
+    auto keepDims = op1.keepdimsAttr(); // ::mlir::IntegerAttr
+
+    Value data = op1.data(); // ONNX operands
+
+    auto ty = IntegerType::get(op1.getContext(), 64);
+
+    // Reading the ONNX side pads values and store in the array.
+    std::vector<Value> axesList = createArrayAttribute(axes, ty, loc, rewriter);
+
+    auto zero = 0;
+    auto f00 = IntegerAttr::get(ty, zero);
+    auto f0 = f00;
+    Value f0v = rewriter.create<ConstantIntOp>(loc, f0);
+
+    Value keepdimVal;
+    if (keepDims)
+      keepdimVal = rewriter.create<ConstantIntOp>(loc, keepDims);
+    else
+      keepdimVal = f0v;
+
+    Value axesShapeList = rewriter.create<PrimListConstructOp>(loc,
+        Torch::ListType::get(rewriter.getType<Torch::IntType>()),
+        ValueRange{axesList});
+
+    TensorType data_tensor_type = data.getType().cast<TensorType>();
+    TensorType op_tensor_type = op->getResult(0).getType().cast<TensorType>();
+
+    auto dataTy = Torch::ValueTensorType::get(context,
+        data_tensor_type.getShape(), data_tensor_type.getElementType());
+    auto dtt = rewriter.create<torch::TorchConversion::FromBuiltinTensorOp>(
+        loc, dataTy, data);
+    auto resultTy = Torch::ValueTensorType::get(op1.getContext(),
+        op_tensor_type.getShape(), op_tensor_type.getElementType());
+
+    llvm::outs() << "axesShapeList:"
+                 << "\n"
+                 << axesShapeList << "\n"
+                 << "\n";
+    llvm::outs() << "keepdimVal:"
+                 << "\n"
+                 << keepdimVal << "\n"
+                 << "\n";
+    llvm::outs() << "dtt Value:"
+                 << "\n"
+                 << dtt << "\n"
+                 << "\n";
+
+    Value atenmultensorOp =
+        rewriter.create<AtenMulTensorOp>(loc, resultTy, dtt, axesShapeList);
+    // Value atenmultensorOp = rewriter.create<AtenMulTensorOp>(loc, resultTy,
```

Comment:
@ramakota-helprack 

I'm not sure that this AtenMulTensorOp is the equivalent of the ReduceMean operation on the ONNX side.  Can you please check if this is a complete and proper implementation?

Reply:
Resolved